pipeline_utils/pipeline_deploy.py

```diff
@@ -0,0 +1,406 @@
+from pipeline_utils.lib import yaml_parser
+
+
+###############################################################
+#   REPOSITORY
+#
+### Expected structure for the pipeline repository
+#
+#   pipeline
+#       wfl
+#       dockerfiles
+#            <image>
+#                Dockerfile
+#       portal_objects
+#           workflows
+#               <wfl>.yaml
+#           metaworkflows
+#               <mwfl>.yaml
+#           file_format.yaml
+#           file_reference.yaml
+#           software.yaml
+#       PIPELINE
+#       VERSION
+#
+###############################################################
+
+
+###############################################################
+#   Logger
+###############################################################
+logger = structlog.getLogger(__name__)
+
+
+###############################################################
+#   PostPatchRepo, class definition
+###############################################################
+class PostPatchRepo(object):
+
+
+    def __init__(self, args, repo, version='VERSION', pipeline='PIPELINE'):
+        """
+        """
+        # Init credentials
+        self.ff_key = None
+        self.kms_key_id = None
+        self.repo = repo
+        self.object_ = {
+            'Software': yaml_parser.YAMLSoftware,
+            'FileFormat': yaml_parser.YAMLFileFormat,
+            'FileReference': yaml_parser.YAMLFileReference,
+            'Workflow': yaml_parser.YAMLWorkflow,
+            'MetaWorkflow': yaml_parser.YAMLMetaWorkflow
+        }
+        self.filepath = {
+            'Software': 'portal_objects/software.yaml',
+            'FileFormat': 'portal_objects/file_format.yaml',
+            'FileReference': 'portal_objects/file_reference.yaml',
+            'Workflow': 'portal_objects/workflows',
+            'MetaWorkflow': 'portal_objects/metaworkflows',
+            'ECR': 'dockerfiles',
+            'WFL': 'wfl'
+        }
+
+        # Load attributes
+        for key, val in vars(args).items():
+            setattr(self, key, val)
+
+        # Get pipeline version
+        with open(f'{self.repo}/{version}') as f:
+            self.version = f.readlines()[0].strip()
+        # Get pipeline name
+        with open(f'{self.repo}/{pipeline}') as f:
+            self.pipeline = f.readlines()[0].strip()
+
+        # Load credentials
+        # self._get_credentials()
+
+
+    def _get_credentials(self):
+        """Get auth credentials to target portal and environment.
+        """
+        # Get portal credentials
+        if os.environ.get('GLOBAL_BUCKET_ENV') and os.environ.get('S3_ENCRYPT_KEY'):
+            # new account
+            s3 = s3_utils.s3Utils(env=self.ff_env)
+            self.ff_key = s3.get_access_keys('access_key_admin')
+        elif os.environ.get('S3_ENCRYPT_KEY'):
+            # main account, also need the key to get auth this way
+            self.ff_key = ff_utils.get_authentication_with_server(ff_env=self.ff_env)
```

Comment:
I think `ff_utils.get_authentication_with_server` is just a roundabout way of doing the same thing as you're doing above, so can probably just do the following? Also, I'm not sure the `S3_ENCRYPT_KEY` is required to do these operations, as not all environments have S3 encryption, so maybe a try/except clause works better. However, the above pertains to current master dcicutils, which is at `v4.4.1`, whereas you're using `v4.0.1`, so consider bring in latest or disregarding the above. @willronchetti may have a better idea on the cleanest way to manage this.
```suggestion
        if os.environ.get('S3_ENCRYPT_KEY'):
            s3 = s3_utils.s3Utils(env=self.ff_env)
            self.ff_key = s3.get_access_keys('access_key_admin')
```

Reply:
I honestly got a bit lost with all the dcicutils updates. I am fine with whatever is currently the right solution to get this auth. @willronchetti can give some feedback on that and I am happy to implement the right solution

Reply:
Couple points worth mentioning:

1. All environments have `S3_ENCRYPT_KEY` - this key is used for the system bucket credentials and is customer managed. You can't get admin credentials or run foursight/tibanna without it.
2. Only some environments have `S3_ENCRYPT_KEY_ID`, which is relevant in some situations here. Any inconsistencies here will cause S3 uploads to fail, such as if you have it set in an unencrypted env, you have the wrong one set, or you do not have it set in an encrypted env.
3. `GLOBAL_ENV_BUCKET` should be set always for internal operations. Without this `dcicutils` cannot get a foothold on the environment structure and will have nowhere to go.

Now, for the task at hand, typically we do not assume presence of the encryption key, as we don't want to assume folks will have this. The right thing to do is rely on the KeyManager interface in dcicutils to grab server and key information from the keydicts file: https://github.com/4dn-dcic/utils/blob/master/dcicutils/creds_utils.py#L286. Recall that the keydicts file has structure like this:

```
{                                                                                                                                                                                                                                                                            
    "cgap-devtest": {
        "key": <access key id>,
        "secret": <access key secret>,
        "server": "https://cgap-devtest.hms.harvard.edu"
    }, ... more envs
}
```

So there are two ways to get a foothold in the new paradigm - `GLOBAL_ENV_BUCKET` for internal users with S3 access and the keys file (which maps environment names to servers along with the needed credentials). Generally you want to rely on the latter. However, because these are workflows you require S3 access to the CWL bucket and ECR access to push the actual workflow images. So it is likely the caller has sufficiently permissioned AWS credentials to use the former method as well.

All that said, the following I believe should work in all situations now. Note that `GLOBAL_BUCKET_ENV` has transitioned to `GLOBAL_ENV_BUCKET`.

```
if os.environ.get('GLOBAL_ENV_BUCKET') and os.environ.get('S3_ENCRYPT_KEY'):
    s3 = s3_utils.s3Utils(env=self.ff_env)
    self.ff_key = s3.get_access_keys('access_key_admin')
elif os.path.exists(self.keydicts_json):
    ... do keydicts setup you're doing
else:    
    raise Exception('Required deployment vars not set: GLOBAL_BUCKET_ENV and S3_ENCRYPT_KEY and no env entry exists in keydicts file!')
```

We will need to update this in the future though to allow non-internal users to add pipelines. Our reliance on ECR creates some difficulty here as we will need to figure how to properly federate access to allow that, or provide an environment that gives that permission. The CWL bucket access is simpler and can be down with the standard pre-signed URLs. It's possible we can use the same mechanism for ECR, just haven't tried it.

Hope this helps clear things up on what's going on!